clients/drcachesim/tracer/tracer.cpp

```diff
@@ -1823,7 +1846,21 @@ event_thread_exit(void *drcontext)
                        "ERROR: The last recorded syscall %d of thread T%d wasn't be "
                        "stopped.\n",
                        cur_recording_sysnum, dr_get_thread_id(drcontext));
-                ASSERT(cur_recording_sysnum, "syscall recording is not stopped");
+
+                /* Write a marker to userspace raw trace. Some drmemtrace derivations
+                 * may interleave the PT trace raw data with the drmemtrace user-space raw
+                 * trace data (instead of outputting the PT trace data to separate files
+                 * like we do here). In such cases, we want to ensure that the
+                 * TRACE_MARKER_TYPE_SYSCALL_IDX does not get output before the actual
+                 * PT trace data, so we output the marker when we stop and write the PT
+                 * trace (instead of when we start the PT trace).
+                 * Note that the order below does not matter because the actual buffer
+                 * flush happens later.
+                 */
+                trace_marker_type_t marker_type = TRACE_MARKER_TYPE_SYSCALL_IDX;
+                uintptr_t marker_val = data->syscall_pt_trace.get_traced_syscall_idx();
+                BUF_PTR(data->seg_base) += instru->append_marker(BUF_PTR(data->seg_base),
+                                                                 marker_type, marker_val);
```

Comment:
Please make a helper function to avoid duplicating this code and its long comment in 3 different places.

Reply:
Done.